### PR TITLE
builtins: add builtin to retrieve the payload(s) for a span.

### PR DESCRIPTION
**Pull request by angelapwen**
Part of addressing #55733 

The `payloads_for_span` builtin retrieves all payloads for
a given span ID, given that the span is part of an active trace.
The payloads are returned in JSONB format. If the span is not
found, or if the span does not have any payloads, the builtin
returns an empty JSON object.

With the appropriate usage of this builtin and the
`crdb_internal.trace_id` builtin as shown in the `contention_event`
logic test, all payloads for the current trace may be surfaced.

Release note (sql change): add `payloads_for_span` builtin that
takes in a span ID and returns its paylods in JSONB format. If
the span is not found, or if the span does not have any payloads,
the builtin returns an empty JSON object.

### CODE REVIEW

File: pkg/sql/sem/builtins/builtins.go
Comment (irfansharif) [inline pkg/sql/sem/builtins/builtins.go:3611]: Is there a reason we need to let the scope for `payload` and `err` extend outside the Structured closure below?

File: pkg/sql/sem/builtins/builtins.go
Comment (irfansharif) [inline pkg/sql/sem/builtins/builtins.go:3625]: We don't need to be collecting to `payloads`, do we?

File: pkg/sql/sem/builtins/builtins.go
Comment (irfansharif) [inline pkg/sql/sem/builtins/builtins.go:3629]: Don't we want to return iterutil.StopIteration() instead? This will continue iterating I believe.

File: pkg/sql/sem/builtins/builtins.go
Comment (irfansharif) [inline pkg/sql/sem/builtins/builtins.go:3634]: Will this ever evaluate to being true? We're instantiating `builder` unconditionally above?

**Review by irfansharif — COMMENTED** [review]
(Holding off on further review until we introduce a tabular for for span payloads)

File: pkg/sql/sem/builtins/builtins.go
Comment (angelapwen) [reply pkg/sql/sem/builtins/builtins.go:3611]: Not anymore! Thanks, removing locally.

File: pkg/sql/sem/builtins/builtins.go
Comment (angelapwen) [reply pkg/sql/sem/builtins/builtins.go:3625]: Yep, you're right, this was a vestige of the string version we had played around with! Removing.

File: pkg/sql/sem/builtins/builtins.go
Comment (angelapwen) [reply pkg/sql/sem/builtins/builtins.go:3634]: It doesn't ever evaluate to true — this is why I had never hit the panic when `NewDJSON` is passed `nil` yesterday. I'm removing this and in the next version where I get the span from its ID in the map, will return an empty JSON array if the span ID isn't found.

File: pkg/sql/sem/builtins/builtins.go
Comment (angelapwen) [reply pkg/sql/sem/builtins/builtins.go:3629]: You're right, though now that I am not going to use the `VisitSpans` method (because I can use the `activeSpans` map) this is no longer relevant.

**Comment by angelapwen** [comment]
I need to rebase on top of #60666 and then push up my newest changes that use the `activeSpans` map rather than `VisitSpans()`. Please hold 😸

**Comment by angelapwen** [comment]
Ok, have rebased and resolved comments 🙆‍♀️ if #59992 is merged soon I'm inclined to fix up the commits here and merge what we have, then work on improving the output of the built-in to return a table separately. Otherwise I'll also do that in this PR.

**Comment by irfansharif** [comment]
I'd personally prefer to iterate right here, especially if we're considering changing the output form anyway. I imagine it'll end up undoing most of what was introduced in the last commit.

**Comment by angelapwen** [comment]
@irfansharif I'm not fully understanding why changing the output to table form will require the iteration over all spans. My understanding is that moving output to table form will display one row for each structured payload _in a given span_ so the only iteration we should need to do is through this span's recording(s), which is still happening in the last commit.

**Comment by angelapwen** [comment]
I've now rebased off of master now that #59992 is merged ⚡ and removed the `num_payloads` column of the `node_inflight_trace_spans` vtable that #59992 introduced.

**Review by knz — APPROVED** [review]
This PR is OK as a transition point. Ok to merge this after you spell out what's going to happen next.

Another confirmation that the JSON array approach is defective is that it does not perform memory monitoring and so the accumulation in RAM can cause the node to crash with an OOM error.

This would be alleviated by a streaming approach, or adding memory monitoring to the current implementation).

> Reviewed 15 of 15 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60616#-:-MToge4L1GzoKn59vArg:b-nlspj5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

**Comment by angelapwen** [comment]
Thanks @knz :sparkle: To make sure I understand correctly, when I make the modification so that the built-in returns a table, there should not be an OOM error because the value generator functions stream values into rows default? 

As for what's going to happen next:
- Update the `payloads_for_spans` builtin to return not a JSONB Array but instead a table using the generator functions. Each row of the table would represent a single payload attached to the span. This table should include columns including at least `type` and `JSON` so that a user would be able to filter off of a certain type (or other attribute). 
- After this, add a `pretty` column to the table returned and have the JSON pretty-printed as the value.

**Review by angelapwen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/sql/sem/builtins/builtins.go, line 3611 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/60616#-MTfNgk-nfrlpTWe2JUi-r11-3611:-MTokety-T8dVLlCB0eT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18fbd2dbdfb344498c81c7850d631c761004e0ce/pkg/sql/sem/builtins/builtins.go#L3611)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Not anymore! Thanks, removing locally.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 3625 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/60616#-MTfNgk-nfrlpTWe2JUi-r11-3625:-MTokhMn9piV1EViwpnr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18fbd2dbdfb344498c81c7850d631c761004e0ce/pkg/sql/sem/builtins/builtins.go#L3625)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Yep, you're right, this was a vestige of the string version we had played around with! Removing.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 3629 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/60616#-MTfNgk-nfrlpTWe2JUi-r11-3629:-MTokj8l-kt97P4Di3PZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18fbd2dbdfb344498c81c7850d631c761004e0ce/pkg/sql/sem/builtins/builtins.go#L3629)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

You're right, though now that I am not going to use the `VisitSpans` method (because I can use the `activeSpans` map) this is no longer relevant. 
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 3634 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/60616#-MTfNgk-nfrlpTWe2JUi-r11-3634:-MTokiPE2eBFMJHXNSud:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18fbd2dbdfb344498c81c7850d631c761004e0ce/pkg/sql/sem/builtins/builtins.go#L3634)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

It doesn't ever evaluate to true — this is why I had never hit the panic when `NewDJSON` is passed `nil` yesterday. I'm removing this and in the next version where I get the span from its ID in the map, will return an empty JSON array if the span ID isn't found.
</blockquote></details>

Done.

**Review by knz — COMMENTED** [review]
> add a pretty column to the table returned and have the JSON pretty-printed as the value.

thanks, can you explain this point more?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60616#-:-MTpBWaZ2IufWYRSmi8i:brgzlrc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

**Comment by angelapwen** [comment]
@knz I meant that the built-in's table returned could also have a column called `pretty` whose value is the pretty-printed version of the JSON blob (in text format). We would want both this column and the column that returns the actual JSONB value so that a user could also use the JSONB functions on this column.

By the way, I was thinking — should the builtins for `trace_id` and `payloads_for_span` require that a user is admin? If so I can add those checks to this PR.

File: pkg/sql/logictest/testdata/logic_test/contention_event
Comment (irfansharif) [inline pkg/sql/logictest/testdata/logic_test/contention_event:54]: A bit of indentation would help: 

```
WITH spans AS (
  SELECT span_id FROM crdb_internal.node_inflight_trace_spans
  WHERE trace_id = crdb_internal.trace_id()
), payload_types AS (
  SELECT jsonb_array_elements(crdb_internal.payloads_for_span(span_id))->>'@type' AS payload_type,
  FROM spans
) 
SELECT count(*) > 0 FROM payload_types
WHERE payload_type = 'type.googleapis.com/cockroach.roachpb.ContentionEvent';
```

Also avoids `type`, which is an [unreserved keyword](https://www.cockroachlabs.com/docs/v20.2/sql-grammar#unreserved_keyword). Also I'm not sure that casing matters for table names, and AFAIK (not a sql expert!), it's more common to use lowercase forms.

**Comment by irfansharif** [comment]
> By the way, I was thinking — should the builtins for trace_id and payloads_for_span require that a user is admin? If so I can add those checks to this PR.

Yup.

File: pkg/sql/logictest/testdata/logic_test/contention_event
Comment (angelapwen) [reply pkg/sql/logictest/testdata/logic_test/contention_event:54]: Yep, this makes sense! Making these changes now. And you are right about lowercase forms for the table names here.

**Comment by angelapwen** [comment]
Added admin checks for the `trace_id` and `payloads_for_span` builtins 😺

**Review by irfansharif — APPROVED** [review]

**Comment by angelapwen** [comment]
bors r=knz,irfansharif

Thanks for the reviews 😄

**Pull request merged**
